### PR TITLE
feature: add filtering of instructions on the exec command

### PR DESCRIPTION
**Pull request by GregoryEssertel**
## Summary:

 - Add a `--exclude/-e` and `--include/-i` flags on the exec command to filter instructions. The filters are being validated then forwarded to SCLE where the filtering is done: https://github.com/strateos/lab/pull/1527
 - Minor: use email address to specify who sent the payload.

Require: https://github.com/strateos/lab/pull/1526 to be merged in lab as well
### Usage:
```
cat autoprotocol.json | \
transcriptic exec -a lab.core.strateos.com/haven/wctest -w wc4-mcx1 -e 1-9 -i 7 -e op:provision
```
This command will first filter out instruction from index 1 to 9 (`-e 1-9`) and the provisions (`-e op:provision`). Then add back the instruction at index 7 (`-i 7`).

The order of `-i/-e` don't matter, it is first removing all the `-e`, then adding back all the `-i`.

### Filters

Possible filter:
 - single index: 123
 - range: 123-456
 - key:value in the instruction: `op:provision`, or `x_human!:false`. The ! means that if the key (x_human) is not in the instruction, then the filter matches. Otherwise, without the '!' a missing key means that there is no match.

Filter example:
  - { "op": "seal", "x_human": false } and { "op": "seal" } match `x_human!:false`
  - { "op": "seal" } does not match`x_human:true`, nor `x_human:false`

### Test plan
Add unit tests

#### Manual testing

```
internal_protocols/clients/sd2/current_protocols/growth_curve [master*]$ transcriptic preview GrowthCurve | \
transcriptic exec -a lab.core.strateos.com/haven/wctest -w wc4-mcx1 -e x_human:true
Sending request to http://13.52.223.111:9000/testRun
Success. View http://lab.core.strateos.com/haven/wctest/dashboard to see the scheduling outcome.
```
or
```
internal_protocols/clients/sd2/current_protocols/growth_curve [master*]$ transcriptic preview GrowthCurve | \
transcriptic exec -a lab.core.strateos.com/haven/wctest -w wc4-mcx1 -e op:provision
Sending request to http://13.52.223.111:9000/testRun
Success. View http://lab.core.strateos.com/haven/wctest/dashboard to see the scheduling outcome.
```

image

Bonus with the newest features not yet merged:
  - UI upgrade:  https://github.com/strateos/lab-frontend/pull/46
  - Rescheduling with new starting times: https://github.com/strateos/lab/pull/1521
  - Extract device set from file added to the docker container: https://github.com/strateos/lab/pull/1526

image

### CODE REVIEW

**Comment by matthewkuangga** [comment]
Works perfectly on local env. Include and exclude works just as intended. Great new feature that can go hand in hand with the new APR error message index feature

**Review by EribertoLopez — APPROVED** [review]
LGTM!

**Review by matthewkuangga — APPROVED** [review]
LGTM, works perfectly when I tested

File: transcriptic/commands.py [hunk 1]
```diff
@@ -1476,6 +1476,28 @@ def run_protocol(api, manifest, protocol, inputs, view=False, dye_test=False):
...
+    invalid_filters = set()
+
+    for arg in filters:
+        if arg.isdigit():
+            idx = int(arg)
+            if 0 > idx or idx >= number_of_instructions:
+                invalid_filters.add(arg)
+        elif re.match(r"\d+-\d+", arg):
+            [s, e] = [int(v) for v in arg.split("-")]
+            if s > e or number_of_instructions <= e:
```
Comment (scottyantipa) [inline transcriptic/commands.py:1489]: You should also check here that s is > 0

**Review by scottyantipa — APPROVED** [review]
This looks good. I think one nice addition would be to display the result of the filtering to the user, like:
```
Successfully submitted 10/13 instructions
```
To re-inforce to the user that their filter was applied as expected

**Comment by GregoryEssertel** [comment]
@scottyantipa I didn't see your comments :grimacing: 

I will address them in another PRs. At first I did the filtering client side, and I was giving this info. But then I move it server side, I will update the message return by SCLE with this information!

File: transcriptic/commands.py (hunk 1 above)
Comment (GregoryEssertel) [reply transcriptic/commands.py:1489]: I thought about it, I think '\d+' should enforce that right? 
With not 0?

**Pull request merged**